pump/server_test.go

```diff
@@ -555,3 +565,222 @@ func (s *listenSuite) TestReturnListener(c *C) {
+	cfg := &Config{
+		ListenAddr:        "http://127.0.0.1:8250",
+		AdvertiseAddr:     "http://127.0.0.1:8260",
+		Socket:            "unix://127.0.0.1:8280/hello/world",
+		EtcdURLs:          strings.Join(etcdClient.Endpoints(), ","),
+		DataDir:           "/tmp/pump",
+		HeartbeatInterval: 1500,
+		LogLevel:          "debug",
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	grpcOpts := []grpc.ServerOption{grpc.MaxRecvMsgSize(GlobalConfig.maxMsgSize)}
+	p := &Server{
+		dataDir:    cfg.DataDir,
+		storage:    &noOpStorage{},
+		clusterID:  8012,
+		node:       &startNode{status: &node.Status{State: node.Online, NodeID: "startnode-long", Addr: "http://192.168.199.100:8260"}},
+		tcpAddr:    cfg.ListenAddr,
+		unixAddr:   cfg.Socket,
+		gs:         grpc.NewServer(grpcOpts...),
+		ctx:        ctx,
+		cancel:     cancel,
+		tiStore:    nil,
+		gcDuration: time.Duration(cfg.GC) * 24 * time.Hour,
+		pdCli:      nil,
+		cfg:        cfg,
+		triggerGC:  make(chan time.Time)}
+	defer p.Close()
+	go p.Start()
+
+	// wait until the server is online
+	timeEnd := time.After(5 * time.Second)
+	getInterval := time.Tick(500 * time.Microsecond)
+WAIT:
+	for {
+		select {
+		case <-getInterval:
+			resp, err := http.Get("http://127.0.0.1:8250/status")
+			if err != nil {
+				if !strings.Contains(err.Error(), "connect: connection refused") {
+					c.Fatal("Connect to pump server fail, error is " + err.Error())
+				}
+				continue
+			}
+			// should receive valid and correct node status info
+			c.Assert(resp, NotNil)
+			defer resp.Body.Close()
+			bodyByte, err := ioutil.ReadAll(resp.Body)
+			c.Assert(err, IsNil)
+
+			nodesStatus := &HTTPStatus{}
+			err = json.Unmarshal(bodyByte, nodesStatus)
+			c.Assert(err, IsNil)
+			c.Assert(nodesStatus.ErrMsg, Equals, "")
+			c.Assert(nodesStatus.CommitTS, Equals, int64(42))
+			fakeNodeStatus, ok := nodesStatus.StatusMap["startnode-long"]
+			c.Assert(ok, IsTrue)
+			c.Assert(fakeNodeStatus.NodeID, Equals, "startnode-long")
+			c.Assert(fakeNodeStatus.Addr, Equals, "http://192.168.199.100:8260")
+			c.Assert(fakeNodeStatus.State, Equals, node.Online)
+			break WAIT
+		case <-timeEnd:
+			c.Fatal("Wait pump to be online for more than 5 seconds")
+		}
+	}
+
+	// test AllDrainer
+	resultStr := httpRequest(c, http.MethodGet, "http://127.0.0.1:8250/drainers")
+	c.Assert(strings.Contains(resultStr, "can't provide service"), IsTrue)
+	// test BinlogByTS
+	resultStr = httpRequest(c, http.MethodGet, "http://127.0.0.1:8250/debug/binlog/2")
+	c.Assert(strings.Contains(resultStr, "server_test"), IsTrue)
+	// test triggerGC
+	resultStr = httpRequest(c, http.MethodPost, "http://127.0.0.1:8250/debug/gc/trigger")
+	c.Assert(strings.Contains(resultStr, "trigger gc success"), IsTrue)
+	// test pause node
+	resultStr = httpRequest(c, http.MethodPut, "http://127.0.0.1:8250/state/startnode-long/pause")
+	c.Assert(strings.Contains(resultStr, "success"), IsTrue)
+
+	// wait server 1s for stopping
+	time.Sleep(time.Second)
```

Comment:
Check if the server is closed periodically so that we don't have to wait for 1 second every time?

Reply:
I add a `chan` in storage to notify the main function that the closing job is over. Actually `p.isclosed` turns to 1 doesn't mean the `p.Close()` is all over. Now we don't have wait for 1 second any more.